_posts/2022-09-08-patterns-with-suspense.markdown

```diff
@@ -0,0 +1,594 @@
+notably the structure of the component tree and the nesting of various Suspense
+boundaries. Breaking it down a bit:
+
+1. `<App />` is wrapped in a top-level `<Suspense />` boundary. This tells React
+   that it is waiting for something to happen
+1. The first component nested within the top-level suspense boundary is
+   `<UserInfo />`. React attempts to render it
+1. Inside of `<UserInfo />` there's a mechanism for fetching user data. When
+   triggered, this tells React that we need to "suspend"
+1. At this moment the top-level "Loading..." fallback indicator appears.
+1. `<FollowedArtists />` and `<FollowedArtworks />` execute and also trigger
+   fallback loading states
+1. `<UserInfo />` finishes loading and reveals the page
+1. `<FollowedArtists />` and `<FollowedArtworks />` render and show their
+   content as soon as they finish loading
+
+The nested suspense boundaries give us fine-grained control over time management
+in our UI. We're able to say: before the user data is fetched, visually pause
+everything (as user info is critical); once that's done render artists and
+artworks as they are secondary, and the order in which they resolve is
+unimportant. The structure of the boundaries are like adding breakpoints to ones
+code, and some breakpoints are clearly more important than others judging by the
+depth in which the Suspense boundaries appear in the component tree.
+
+The question is: how does a component tell React that it needs to suspend?
+
+## Building your first Suspense component
+
+This is where things get interesting and kind of weird and we discover that what
+seems complex is actually deceptively simple. To instruct React to suspend, you
+simply `throw` a promise in the same way that you would `throw` an error:
+
+```tsx
+const UserInfo = () => {
+  throw new Promise(...)
+
+  return (
+    <div>...</div>
+  )
+}
+```
+
+That's Suspense's mechanism in a nutshell and all you really need to know to
+understand some of the more complex ideas that follow. In a React component
+render cycle, `throw promise` means: suspend.
+
+## But why?
+
+When I first started exploring Suspense I asked myself this question a lot,
+because I was only thinking about suspense as a way to fetch data, and I had
+already been fetching data with sequential loading states just fine for years.
+Why is it necessary to introduce a strange new JavaScript pattern ("throwing a
+promise?") for something that I can easily do in a `useEffect`?
+
+What I didn't understand is that Suspense is designed for something other than
+_fetching on render_; it's designed around the concept of _rendering as you
+fetch_. This is _the_ crucial idea to understand about Suspense: that **one
+should be able to execute an entire component tree in one go with nothing
+preventing any other component from executing, regardless of whether there are
+required data dependencies needed for rendering the display**. The performance
```

Comment:
I acknowledge I may be a bit jaded, but I have an almost knee-jerk reaction to performance claims these days and just wondering if we have anything to prove this out...?

Reply:
The best way to understand it is to imagine component `<App />`, which contains components <A />, which contains <B />, which contains <C />, and each have their own data dependencies. The "suspense" way is execute `<App />` which in turn fires off A, B and C concurrently. A's request might take a long time, and so we "suspend" it (and show a loading state), wheres B and C are faster. Rather than having to depend on A in order to trigger B and C -- thus creating a waterfall (imagine a crome network panel flamegraph) -- each of those requests are fired in parallel, thus speeding up the entire app. It's a remarkably faster pattern. And remember that "request" can mean anything that takes some measure of time -- a long running computation, for example.

Reply:
(I should probably add this as a paragraph)